File: examples/c/quickstart.c
#include "perbacco.h"

#include <inttypes.h>
#include <stddef.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>

static int fail(pb_engine *engine, pb_status status) {
    const char *detail = pb_engine_last_error(engine);
    fprintf(stderr, "pERbacco: %s\n", detail != NULL ? detail : pb_strerror(status));
    pb_engine_free(engine);
    return EXIT_FAILURE;
}

int main(void) {
    pb_edge edges[] = {
        {0u, 1u, 0.96},
        {2u, 3u, 0.93},
        {0u, 2u, 0.08},
        {1u, 3u, 0.06},
    };
    const uint32_t truth[] = {0u, 0u, 1u, 1u};
    pb_record_id records[3];
    uint32_t labels[3];
    pb_config config;
    pb_engine *engine = NULL;
    pb_stats stats;
    pb_batch batch;

    pb_config_init(&config);
    config.method = PB_METHOD_PERBAC;
    config.cda = PB_CDA_NONE;
    config.batch_size = 3u;
    config.max_queries = 10u;

    pb_status status = pb_engine_create(
        4u,
        edges,
        sizeof(edges) / sizeof(edges[0]),
        &config,
        NULL,
        truth,
        &engine);
    if (status != PB_OK) {
        return fail(engine, status);
    }

    for (;;) {
        status = pb_engine_next_batch(engine, records, 3u, &batch);
        if (status == PB_DONE || status == PB_BUDGET_EXHAUSTED) {
            break;
        }
        if (status != PB_OK) {
            return fail(engine, status);
        }
        for (size_t index = 0u; index < batch.len; ++index) {
            labels[index] = truth[batch.records[index]];
        }
        status = pb_engine_submit_partition(engine, labels, batch.len);
        if (status != PB_OK) {
            return fail(engine, status);
        }
    }

    status = pb_engine_get_stats(engine, &stats);
    if (status != PB_OK) {
        return fail(engine, status);
    }
    printf("queries=%" PRIu64 ", matches=%" PRIu64 "\n",
           stats.query_count,
           stats.discovered_matches);
    pb_engine_free(engine);
    return EXIT_SUCCESS;
}

File: examples/python/quickstart.py
"""Five-minute pERbacco run with a local ground-truth oracle."""

from perbacco import CDA, Engine, EngineConfig, Graph, GroundTruthOracle, Method, run


def main() -> None:
    truth = {"a1": "alice", "a2": "alice", "b1": "bob", "b2": "bob"}
    records = {
        "a1": {"name": "Alice Rossi", "city": "Rome"},
        "a2": {"name": "A. Rossi", "city": "Roma"},
        "b1": {"name": "Bob Smith", "city": "London"},
        "b2": {"name": "Robert Smith", "city": "London"},
    }
    graph = Graph.from_edges(
        [
            ("a1", "a2", 0.96),
            ("b1", "b2", 0.93),
            ("a1", "b1", 0.08),
            ("a2", "b2", 0.06),
        ],
        records=records,
    )
    config = EngineConfig(
        method=Method.PERBACCO,
        cda=CDA.LOUVAIN,
        batch_size=3,
        max_queries=10,
    )

    with Engine(graph, config) as engine:
        result = run(
            engine,
            GroundTruthOracle(truth),
            total_truth_matches=2,
        )

    print(f"queries={result.stats.query_count}")
    print(f"matches={result.stats.discovered_matches}")
    print(f"recall={result.events[-1].recall:.1f}")


if __name__ == "__main__":
    main()
